This folder contains scripts with the specific parameters used for "Leveraging phylogenetic signal to unravel microbial community function and assembly rules" by Talavera-Marcos, Parras-Moltó and Aguirre de Cárcer (https://doi.org/10.21203/rs.3.rs-2272005/v1), plus the original annotation files and the output consensus lists.

### Script
Edited version of `1__core_functions_full.py`, specific for this dataset

### Results
We have have the core pangenome for all nodes found in the glucose, leucine and citrate cultures. Node35562 is common for citrate and glucose. We define "core pangenome" as the list of functions present in 90% of the individual genomes considered.

We also include all the individual annotations used to get the resulting lists. As explained in the manuscript, we only annotated those genomes that we could assign to PCG members above a certain 16S sequence identity threshold.

The obtained lists of annotations can be visualized with the online KEGG mapper tool.
